lib/fetch/request.js

```diff
@@ -419,16 +419,12 @@ class Request {
 
       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
-      if (headers instanceof Headers) {
-        // TODO (fix): Why doesn't this work?
-        // for (const [key, val] of headers[kHeadersList]) {
-        //   this[kHeaders].append(key, val)
-        // }
-
-        this[kState].headersList = new HeadersList([
-          ...this[kState].headersList,
-          ...headers[kHeadersList]
-        ])
+      if (headers.constructor.name === 'Headers') {
+        for (const [key, val] of headers) {
+          this[kHeaders].append(key, val)
+        }
+
+        this[kState].headersList = headers[kHeadersList]
```

Comment:
Why is this needed?

Reply:
node-fetch tests without it 🤷 
see: https://github.com/nodejs/undici/runs/6262358645?check_suite_focus=true

Reply:
This is weird. I'm not comfortable landing before we know what's going on here.

Reply:
So this is why... I think...

this[kState]'s headersList is set to this[kHeaders][kHeadersList] so when appending to 1, both are being appended to. So in the diff below I separate them and fill each separately. That fixes the issue, however I noticed that the spec doesn't say to have 2 different headersList so I think that's part of the confusion.

```diff
diff --git a/lib/fetch/request.js b/lib/fetch/request.js
index 9993b81..6478eda 100644
--- a/lib/fetch/request.js
+++ b/lib/fetch/request.js
@@ -415,19 +415,19 @@ class Request {

       // 3. Empty this’s headers’s header list.
       this[kState].headersList = new HeadersList()
-      this[kHeaders][kHeadersList] = this[kState].headersList
+      this[kHeaders][kHeadersList] = new HeadersList()

       // 4. If headers is a Headers object, then for each header in its header
       // list, append header’s name/header’s value to this’s headers.
       if (headers.constructor.name === 'Headers') {
         for (const [key, val] of headers) {
           this[kHeaders].append(key, val)
+          this[kState].headersList.append(key, val)
         }
-
-        this[kState].headersList = headers[kHeadersList]
       } else {
         // 5. Otherwise, fill this’s headers with headers.
         fillHeaders(this[kState].headersList, headers)
+        fillHeaders(this[kHeaders][kHeadersList], headers)
       }
     }
```

Reply:
I don't know if this is an acceptable fix because it seems slower than just letting it be weird lol

Reply:
I found the problem.